source/iterator/cdc.go

```diff
@@ -0,0 +1,239 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/miquido/conduit-connector-azure-storage/internal"
+	"github.com/miquido/conduit-connector-azure-storage/source/position"
+	"gopkg.in/tomb.v2"
+)
+
+var ErrCDCIteratorIsStopped = errors.New("CDC iterator is stopped")
+
+func NewCDCIterator(
+	pollingPeriod time.Duration,
+	client *azblob.ContainerClient,
+	from time.Time,
+	maxResults int32,
+) (*CDCIterator, error) {
+	if maxResults < 1 {
+		return nil, fmt.Errorf("maxResults is expected to be greater than or equal to 1, got %d", maxResults)
+	}
+
+	cdc := CDCIterator{
+		client:        client,
+		buffer:        make(chan sdk.Record, 1),
+		ticker:        time.NewTicker(pollingPeriod),
+		isTruncated:   true,
+		nextKeyMarker: nil,
+		tomb:          tomb.Tomb{},
+		lastModified:  from,
+		maxResults:    maxResults,
+	}
+
+	cdc.tomb.Go(cdc.producer)
+
+	return &cdc, nil
+}
+
+type CDCIterator struct {
+	client        *azblob.ContainerClient
+	buffer        chan sdk.Record
+	ticker        *time.Ticker
+	lastModified  time.Time
+	maxResults    int32
+	isTruncated   bool
+	nextKeyMarker *string
+	tomb          tomb.Tomb
+}
+
+func (w *CDCIterator) HasNext(_ context.Context) bool {
+	return len(w.buffer) > 0 || !w.tomb.Alive() // if tomb is dead we return true so caller will fetch error with Next
+}
+
+func (w *CDCIterator) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case r, active := <-w.buffer:
+		if !active {
+			return sdk.Record{}, ErrCDCIteratorIsStopped
+		}
+
+		return r, nil
+
+	case <-w.tomb.Dead():
+		return sdk.Record{}, w.tomb.Err()
+
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err()
+	}
+}
+
+func (w *CDCIterator) Stop() {
+	w.ticker.Stop()
+	w.tomb.Kill(ErrCDCIteratorIsStopped)
+}
+
+// producer reads the container and reports all file changes since last time.
+func (w *CDCIterator) producer() error {
+	defer close(w.buffer)
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return w.tomb.Err()
+
+		case <-w.ticker.C:
+			currentLastModified := w.lastModified
+
+			// Prepare the storage iterator
+			blobListPager := w.client.ListBlobsFlat(&azblob.ContainerListBlobsFlatOptions{
+				Marker:     w.nextKeyMarker,
+				MaxResults: &w.maxResults,
+				Include: []azblob.ListBlobsIncludeItem{
+					azblob.ListBlobsIncludeItemDeleted,
+				},
+			})
+
+			for blobListPager.NextPage(w.tomb.Context(context.TODO())) {
```

Comment:
Should we replace context.TODO()? Also, does this mean that every pollingPeriod, we pull all object metadata from Azure Storage?

Reply:
Changed it to `context.Background()`.

Yes, I'm paginating over container blobs with metadata every pollingPeriod.